In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.impute import SimpleImputer

In [3]:
df= pd.read_csv("titanic.csv")

In [4]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [ ]:
# will drop the unessary column like Passenger id, Name, and focus on what are important column which impact the survial rate.

In [14]:
df_select= df[["Survived", "Pclass", "Sex", "Age", "Fare"]]

In [15]:
# will convert sex into numerical data with the help of Map function.
df_select["Sex"]= df_select["Sex"].map({"male": 0, "female": 1})

C:\Users\User\AppData\Local\Temp\ipykernel_9128\2509868708.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_select["Sex"]= df_select["Sex"].map({"male": 0, "female": 1})


In [16]:
#Now df_select head
df_select.head()

,Survived,Pclass,Sex,Age,Fare
0,0,3,0,22.0,7.2500
1,1,1,1,38.0,71.2833
2,1,3,1,26.0,7.9250
3,1,1,1,35.0,53.1000
4,0,3,0,35.0,8.0500


In [17]:
df_select.dtypes

Survived      int64
Pclass        int64
Sex           int64
Age         float64
Fare        float64
dtype: object

In [18]:
df_select.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
Fare          0
dtype: int64

In [19]:
# handle missing value with this method calculte mean and fill with it. IF someone want to do with Imputer, can also be done. 
mean_age = df_select["Age"].mean()
df_select["Age"] = df_select["Age"].fillna(mean_age)

C:\Users\User\AppData\Local\Temp\ipykernel_9128\1590093461.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_select["Age"] = df_select["Age"].fillna(mean_age)


In [20]:
df_select["Age"].isnull().sum()

np.int64(0)

In [21]:
df_select.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
Fare        0
dtype: int64

In [22]:
#Now define X and X
X= df_select.drop(["Survived"], axis= 1)
y= df_select["Survived"]

In [23]:
X.head()

,Pclass,Sex,Age,Fare
0,3,0,22.0,7.2500
1,1,1,38.0,71.2833
2,3,1,26.0,7.9250
3,1,1,35.0,53.1000
4,3,0,35.0,8.0500


In [24]:
y.head()

0    0
1    1
2    1
3    1
4    0
Name: Survived, dtype: int64

In [25]:
# split the data into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [28]:
#now bring GaussainNB
GNB_model= GaussianNB()

In [29]:
GNB_model.fit(X_train, y_train)

GaussianNB()

In [30]:
#now we will do prediction:
y_pred= GNB_model.predict(X_test)

In [31]:
y_test

709    1
439    0
840    0
720    1
39     1
      ..
433    0
773    0
25     1
84     1
10     1
Name: Survived, Length: 179, dtype: int64

In [32]:
#we can compare also with true result and pred result
df_compare = pd.DataFrame({
    "Actual": y_test,
    "Predicted": y_pred
})

print(df_compare.head())

     Actual  Predicted
709       1          0
439       0          0
840       0          0
720       1          1
39        1          1


In [33]:
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print(" Classification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.7597765363128491
Confusion Matrix:
 [[83 22]
 [21 53]]
 Classification Report:
               precision    recall  f1-score   support

           0       0.80      0.79      0.79       105
           1       0.71      0.72      0.71        74

    accuracy                           0.76       179
   macro avg       0.75      0.75      0.75       179
weighted avg       0.76      0.76      0.76       179



#Maximize overall accuracy (currently ~76%)
#Maintain a good balance between precision and recall

#From confusion matrix:

[[83 22]
 [21 53]]
True Negatives (83) → correctly predicted (Survived)
True Positives (53) → correctly predicted (not survived)
False Positives (22) → predicted (not survived but actually survived)
False Negatives (21) → missed result